### ISSUE TITLE
Could lead to memory explosion ?

### ISSUE THREAD

**Issue by Matsa59** [IvanIvanoff/absinthe_cache#9]
According to the code in https://github.com/IvanIvanoff/absinthe_cache/blob/master/lib/document_provider.ex every time the server receive a different GQL request it will save it to the cache.

Every different GQL request it will save it to the cache

Imagine we have a user object

```elixir
object :user do
  field :id, :integer,
  field :name, :string
  field :description :description
end
```

We could perform 7 differents queries (I don't remember the formule but it's like exponential)
- id
- id, name
- id, description
- name
- name, description
- description
- id, name, description

So for one simple object we could have a lot of different cache key or I'm missing something?

Maybe register a "pre-result" could save a lot of memory space (ets). Then apply a transformation on the "pre-result" to generate the result following the user's selections.

**Comment by IvanIvanoff** [comment]
Hi @Matsa59,

Sorry for the late reply, but I have not expected to have people look into this and ask questions, so I have not monitored the issues here.

The short answer is that this `Maybe register a "pre-result" could save a lot of memory space (ets)`  is already done.

Now the long answer.

The cache key depends on the query name and arguments, not the selected fields. If your resolver returns 10 fields then all the fields are computed and returned every time. No matter what is your selection, the resolver will return all fields (and cache them) and only after that some of them will be filtered out. So the following two graphql queries have the same cache key:
```graphql
{
  getUser(id: 5){
    id
  }
}
```
```graphql
{
  getUser(id: 5){
    id
    name
    description
  }
}
```


The more interesting case is what happens when the fields themselves have resolvers themselves:
```graphql
object :user do
  field :id, :integer
  field :name, :string do
     cache_resolve(&UserResolver.get_user_name/3)
  end
  field :description :description
end
```
then the same logic applies (but twice) - you would have 2 different cache keys that depend only on the query name and its arguments. You could actually need to fetch data from the cache only once if the `name` field is not selected.

Now, this is not the case with the DocumentProvider cache. The above approach has one glaring issue - it makes a lot of function calls and calls to ETS which always result in copying data into your process. In my own codebase I have some queries that return a list of 2000+ objects, each of which has 10+ fields that are computed via custom resolvers. This will lead to at least 20k calls to ETS that copy data, which is more than I am willing to allow.

The DocumentProvider cache bypasses the Resolution phase altogether - no `resolve` or `cache_resolve` invocations are made, thus we cannot leverage the fact that the filtering is done after the resolvers return. Because of this, only in this scenario we inspect the arguments, so the whole query can be "computed" (fetched from the cache) with a single ETS call. Only here it matters if the selection has only `id` or `id, name, description`.

Here are my modules that use the DocumentProvider cache: 
1. [DocumentProvider Module](https://github.com/santiment/sanbase2/blob/master/lib/sanbase_web/graphql/document/document_provider.ex)
2. [Absinthe Before Send Module](https://github.com/santiment/sanbase2/blob/master/lib/sanbase_web/graphql/absinthe_before_send.ex)

Sadly, I still cannot allocate enough time to write much better docs (and finally release this as a library to hex) but I would be happy to help you understand better and integrate the code, if you need.

**Comment by Matsa59** [comment]
Hello,

My concerne is more about the size of the cache. I imagine "hackers" that send x differents gql document just for create a lot of elements in the cache. 

I saw a video on youtube made by Facebook about how to optimize GQL. The best approch is to associate a "request_id" to a GQL request. So client only send the `request_id` + `variables` (it reduces requests sizes and don't allow client to define their request but it's useless in case of a public api).

But after thinking about it, I wonder if it could happen since the cache have a short TTL.

Keep going, you made a good job ;)

**Comment by IvanIvanoff** [comment]
I am running the cache at https://app.santiment.net/ with a reasonable amount of traffic (millions of API calls per day). Currently the ETS table size is ~215Mb, so I have not had any issues so fartc. I have experienced much more issues with Absinthe itself, as some single requests were resulting in huge process sizes - I have seen a single GraphQL request take more than 10Mb of memory only due to Absinthe's implementation.

The cache size must be controlled by the cache backend. The cache libraries (`con_cache`, `cachex`, etc.) support different type of configurations. I think the best approach would be to configure the cache to purge according to LRU (least recently used) rules and keep maximum of N key-value pairs. I cannot say from the top of my head what the `con_cache` library supports, but I know for sure that `cachex` definitely supports that.  I have a `cachex` backend written but I have not added it here as I have experienced dead locks with it - after some time the cache just stops responding. I cannot say for sure if it was its fault or my fault but I'm more inclined to believe it is my fault as `cachex` is a pretty popular library. I'll try to add it next week here and add examples with this type of configuration.

The cache backend needs to implement very few callbacks and in almost all cases their implementation is just mapping the required callback to the used library's API. In case the cache is expected to grow very big it is possible to add a Redis cache, which would live on a separate machine with higher memory resources.
